controllers/internal/utils/dns.go

```diff
@@ -20,30 +20,99 @@ package utils
 import (
 	"fmt"
 	"sort"
+	"strings"
 
 	"github.com/lixiangzhong/dnsutil"
+	"github.com/miekg/dns"
 )
 
+type DNSServer struct {
+	Host string
+	Port int
+}
+
 // Dig retrieves list of tuple <IP address, A record > from edge DNS server for specific FQDN
 func Dig(edgeDNSServer, fqdn string) ([]string, error) {
-	var dig dnsutil.Dig
 	if edgeDNSServer == "" {
 		return nil, fmt.Errorf("empty edgeDNSServer")
 	}
-	err := dig.SetDNS(edgeDNSServer)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)", edgeDNSServer, err)
-		return nil, err
+	dnsServer := DNSServer{
+		Host: edgeDNSServer,
+		Port: 53,
+	}
+	return DigWithFallback([]DNSServer{dnsServer}, fqdn)
+}
+
+func DigWithFallback(edgeDNSServers []DNSServer, fqdn string) ([]string, error) {
+	var dig dnsutil.Dig
+	if len(edgeDNSServers) == 0 {
+		return nil, fmt.Errorf("empty edgeDNSServers, provide at least one")
 	}
-	a, err := dig.A(fqdn)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't dig fqdn(%s) with error(%s)", fqdn, err)
-		return nil, err
+	for i := 0; i < len(edgeDNSServers); i += 2 {
+		if edgeDNSServers[i].Host == "" {
```

Comment:
is that valid ? Shouldn't fail here?

Reply:
hmm, I think this should never occur when called from the operator, because the empty dns server should fail the validation phase, but it was there before so I left it there.

The `Dig` helper is called also from tests so I assume the check is there because of it. The goal of the function (as I understand it) is to get the `[]DNSServer` by a) using explicitly selected dns server b) using a list of dns servers in a fallback fashion that if first fails, another tries and whoever will get the results first will return them.

Reply:
This is the question. The point is, the utility `Dig` function is `public static`, which means, it can be potentionally called out of operator universe (internality is irrelevant now). That means, the input should be validated (at least for empty args etc..) and tested.

Reply:
I've added some defensive programming-fu for the Dig itself in https://github.com/k8gb-io/k8gb/commit/ea4c3c950edc6862f310d0bce06b49dad741c8e6#diff-2b429a0ff271b1687026b627ed409f0973f341a47d1f3c4a3f711b42adbc11ccR58 I am still not convinced about failing fast in the `DigWithFallback` in case of one empty sever because it would kind of defeat the purpose of being robust variant of Dig that wouldn't fail if the server is, say, unreachable

But if others also think this is better, I can change that. It's 1 line of code anyway :D

Reply:
I've changed that